r1cs/gadgets/crypto/src/merkle_tree/mod.rs

```diff
@@ -1,17 +1,85 @@
-use algebra::Field;
+use algebra::{Field, PrimeField};
 use r1cs_core::{ConstraintSystem, SynthesisError};
-use r1cs_std:: prelude::*;
+use r1cs_std::prelude::*;
 
-use primitives::{
-    crh::FixedLengthCRH,
-    merkle_tree::*
-};
-use crate::FixedLengthCRHGadget;
+use primitives::{crh::FixedLengthCRH, merkle_tree::*, FieldBasedHash};
+use crate::{FixedLengthCRHGadget, FieldBasedHashGadget};
 
 use std::borrow::Borrow;
+use r1cs_std::fields::fp::FpGadget;
 
 pub mod field_based_mht;
 
+pub trait FieldBasedMerkleTreePathGadget<
```

Comment:
Why do we introduce a generic FieldBasedMerkleTreePathGadget beside the binary one? Do we ever target a different arity?

Reply:
We may in future, yes